### PR TITLE
Add Homeassistant Discovery Feature

### PR DESCRIPTION
**Pull request by dafes**
* Updated MQTT.h to add core and supporting functions
* Updated userConfig_sample.h to introduce new options:
  *  #define MQTT_HASSIO_SUPPORT 0                           // Enables the Homeassistant Auto Discovery Feature
  * #define MQTT_HASSIO_DISCOVERY_PREFIX "homeassistant"    // Homeassistant Auto Discovery Prefix
* Implemented a ToDo inside MQTT and also cleaned it a bit
* Established a Task which runs every n Seconds to update Hassio Discovery defined by `const unsigned long HomeAssistantDiscoveryExecutionInterval = 300000;` in main.cpp

### CODE REVIEW

**Comment by einarhauks** [comment]
Why was the pull request closed without merging into master?

**Comment by dafes** [comment]
The PR is open for 2 Months. @einarhauks  if you tested it please confirm it here so this can finally be merged or closed

**Comment by LoQue90** [comment]
It's only @dafes and me that I know that uses home assistant.
I just had this PR running once, had no time at all besides live, sending out pcbs and giving support.
So any other testing is highly welcome and appreciated!
@einarhauks go for feedback pls
@dafes could you have a look at @339 and maybe add this here.
I'm at home the next days, going to flash ist tomorrow on my main mashine.

**Comment by LoQue90** [comment]
@einarhauks did you had time to test?

**Comment by einarhauks** [comment]
Tested and it looks good. Editing configuration from Home Assistant works as expected and temperature is reported

I suggest improving the texts:
` DiscoveryObject brewSetpoint = GenerateNumberDevice("brewSetpoint", "Brew setpoint", BREW_SETPOINT_MIN, BREW_SETPOINT_MAX, 0.1, "°C");
  DiscoveryObject steamSetPoint = GenerateNumberDevice("steamSetpoint", "Steam setpoint", STEAM_SETPOINT_MIN, STEAM_SETPOINT_MAX, 0.1, "°C");
  DiscoveryObject brewTempOffset = GenerateNumberDevice("brewTempOffset", "Brew Temp. Offset", BREW_TEMP_OFFSET_MIN, BREW_TEMP_OFFSET_MAX, 0.1, "°C");
  DiscoveryObject brewPidDelay = GenerateNumberDevice("brewPidDelay", "Brew Pid Delay", BREW_PID_DELAY_MIN, BREW_PID_DELAY_MAX, 0.1, "");
  DiscoveryObject startKp = GenerateNumberDevice("startKp", "Start kP", PID_KP_START_MIN, PID_KP_START_MAX, 0.1, "");
  DiscoveryObject startTn = GenerateNumberDevice("startTn", "Start Tn", PID_TN_START_MIN, PID_TN_START_MAX, 0.1, "");
  DiscoveryObject steamKp = GenerateNumberDevice("steamKp", "Start Kp", PID_KP_STEAM_MIN, PID_KP_STEAM_MAX, 0.1, "");
  DiscoveryObject aggKp = GenerateNumberDevice("aggKp", "aggKp", PID_KP_REGULAR_MIN, PID_KP_REGULAR_MAX, 0.1, "");
  DiscoveryObject aggTn = GenerateNumberDevice("aggTn", "aggTn", PID_TN_REGULAR_MIN, PID_TN_REGULAR_MAX, 0.1, "");
  DiscoveryObject aggTv = GenerateNumberDevice("aggTv", "aggTv", PID_TV_REGULAR_MIN, PID_TV_REGULAR_MAX, 0.1, "");
  DiscoveryObject aggIMax = GenerateNumberDevice("aggIMax", "aggIMax", PID_I_MAX_REGULAR_MIN, PID_I_MAX_REGULAR_MAX, 0.1, "");
  DiscoveryObject brewtime = GenerateNumberDevice("brewtime", "Brew time", BREW_TIME_MIN, BREW_TIME_MAX, 0.1, "s");
// Sensor Devices
  DiscoveryObject actual_temperature = GenerateSensorDevice("temperature", "Boiler Temperature");
  DiscoveryObject heaterPower = GenerateSensorDevice("heaterPower", "Heater Power", "ms", "power_factor");
// Switch Devices
  DiscoveryObject pidOn = GenerateSwitchDevice("pidON", "Use PID");
  DiscoveryObject steamON = GenerateSwitchDevice("steamON", "Steam");
  DiscoveryObject backflushON = GenerateSwitchDevice("backflushON", "Backflush");
  DiscoveryObject startUsePonM = GenerateSwitchDevice("startUsePonM", "Use PonM");
  `

File: src/MQTT.h [hunk 1]
```diff
@@ -153,41 +205,296 @@ void mqtt_callback(char *topic, byte *data, unsigned int length) {
...
+
+  number_device.payload_json = numberConfigDocBuffer;
+  return number_device;
+}
+
+/**
+ * @brief Send MQTT Homeassistant Discovery Messages
+ * @return 0 if successful, MQTT connection error code if failed to send messages
+ */
+int sendHASSIODiscoveryMsg() {
```
Comment (einarhauks) [inline src/MQTT.h:433]: I suggest improving the texts
` DiscoveryObject brewSetpoint = GenerateNumberDevice("brewSetpoint", "Brew setpoint", BREW_SETPOINT_MIN, BREW_SETPOINT_MAX, 0.1, "°C");
  DiscoveryObject steamSetPoint = GenerateNumberDevice("steamSetpoint", "Steam setpoint", STEAM_SETPOINT_MIN, STEAM_SETPOINT_MAX, 0.1, "°C");
  DiscoveryObject brewTempOffset = GenerateNumberDevice("brewTempOffset", "Brew Temp. Offset", BREW_TEMP_OFFSET_MIN, BREW_TEMP_OFFSET_MAX, 0.1, "°C");
  DiscoveryObject brewPidDelay = GenerateNumberDevice("brewPidDelay", "Brew Pid Delay", BREW_PID_DELAY_MIN, BREW_PID_DELAY_MAX, 0.1, "");
  DiscoveryObject startKp = GenerateNumberDevice("startKp", "Start kP", PID_KP_START_MIN, PID_KP_START_MAX, 0.1, "");
  DiscoveryObject startTn = GenerateNumberDevice("startTn", "Start Tn", PID_TN_START_MIN, PID_TN_START_MAX, 0.1, "");
  DiscoveryObject steamKp = GenerateNumberDevice("steamKp", "Start Kp", PID_KP_STEAM_MIN, PID_KP_STEAM_MAX, 0.1, "");
  DiscoveryObject aggKp = GenerateNumberDevice("aggKp", "aggKp", PID_KP_REGULAR_MIN, PID_KP_REGULAR_MAX, 0.1, "");
  DiscoveryObject aggTn = GenerateNumberDevice("aggTn", "aggTn", PID_TN_REGULAR_MIN, PID_TN_REGULAR_MAX, 0.1, "");
  DiscoveryObject aggTv = GenerateNumberDevice("aggTv", "aggTv", PID_TV_REGULAR_MIN, PID_TV_REGULAR_MAX, 0.1, "");
  DiscoveryObject aggIMax = GenerateNumberDevice("aggIMax", "aggIMax", PID_I_MAX_REGULAR_MIN, PID_I_MAX_REGULAR_MAX, 0.1, "");
  DiscoveryObject brewtime = GenerateNumberDevice("brewtime", "Brew time", BREW_TIME_MIN, BREW_TIME_MAX, 0.1, "s");
// Sensor Devices
  DiscoveryObject actual_temperature = GenerateSensorDevice("temperature", "Boiler Temperature");
  DiscoveryObject heaterPower = GenerateSensorDevice("heaterPower", "Heater Power", "ms", "power_factor");
// Switch Devices
  DiscoveryObject pidOn = GenerateSwitchDevice("pidON", "Use PID");
  DiscoveryObject steamON = GenerateSwitchDevice("steamON", "Steam");
  DiscoveryObject backflushON = GenerateSwitchDevice("backflushON", "Backflush");
  DiscoveryObject startUsePonM = GenerateSwitchDevice("startUsePonM", "Use PonM");
  `

**Review by einarhauks — APPROVED** [review]
Also tested building and uploading to my machine, works as expected

**Comment by einarhauks** [comment]
@LoQue90 I've tested this and verified, now someone with merge review/rights need to take the next step

**Comment by murmeltier08** [comment]
I can merge it, but i do not have tested it.

**Comment by dafes** [comment]
> I can merge it, but i do not have tested it.

worked on my machine :D

**Comment by LoQue90** [comment]
adding the new names from @einarhauks would be nice
add moving the option to enable/disable the feature to the website instead would be nice.
ist running on my mashine for a while

**Comment by dafes** [comment]
> adding the new names from @einarhauks would be nice add moving the option to enable/disable the feature to the website instead of the userconfig would be nice. ist running on my mashine for a while

let us merge the PR first. You or anybody can create a PR with these suggestions.

**Comment by dafes** [comment]
updated
@einarhauks @LoQue90

**Comment by LoQue90** [comment]
then go for merge and let us improve it later!
@murmeltier08

**Pull request merged**